File: app/src/main/java/ru/p3tr0vich/fuel/factories/FragmentFactory.java
package ru.p3tr0vich.fuel.factories;

import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import ru.p3tr0vich.fuel.FragmentAbout;
import ru.p3tr0vich.fuel.FragmentBackup;
import ru.p3tr0vich.fuel.FragmentBase;
import ru.p3tr0vich.fuel.FragmentCalc;
import ru.p3tr0vich.fuel.FragmentChartCost;
import ru.p3tr0vich.fuel.FragmentFueling;
import ru.p3tr0vich.fuel.FragmentPreferences;

public class FragmentFactory {

    public interface Ids {

        @Retention(RetentionPolicy.SOURCE)
        @IntDef({BAD_ID,
                FUELING,
                CALC,
                CHART_COST,
                BACKUP,
                PREFERENCES,
                ABOUT})
        @interface Id {
        }

        int BAD_ID = -1;
        int FUELING = 0;
        int CALC = 1;
        int CHART_COST = 2;
        int BACKUP = 3;
        int PREFERENCES = 4;
        int ABOUT = 5;

        int MAIN = FUELING;
    }

    private FragmentFactory() {
    }

    @NonNull
    public static Fragment getFragmentNewInstance(@Ids.Id int fragmentId, @Nullable Bundle args) {
        Fragment fragment;

        switch (fragmentId) {
            case Ids.CALC:
                fragment = new FragmentCalc();
                break;
            case Ids.CHART_COST:
                fragment = new FragmentChartCost();
                break;
            case Ids.PREFERENCES:
                fragment = new FragmentPreferences();
                break;
            case Ids.BACKUP:
                fragment = new FragmentBackup();
                break;
            case Ids.ABOUT:
                fragment = new FragmentAbout();
                break;
            case Ids.FUELING:
                fragment = new FragmentFueling();
                break;
            case Ids.BAD_ID:
            default:
                throw new IllegalArgumentException("Fragment bad ID");
        }

        return FragmentBase.newInstance(fragmentId, fragment, args);
    }

    @NonNull
    public static Fragment getFragmentNewInstance(@Ids.Id int fragmentId) {
        return getFragmentNewInstance(fragmentId, null);
    }

    @Ids.Id
    public static int intToFragmentId(int id) {
        return id;
    }

    @NonNull
    public static String fragmentIdToTag(@Ids.Id int id) {
        switch (id) {
            case Ids.ABOUT:
            case Ids.BACKUP:
            case Ids.CALC:
            case Ids.CHART_COST:
            case Ids.FUELING:
            case Ids.PREFERENCES:
                return "TAG_" + String.valueOf(id);
            case Ids.BAD_ID:
            default:
                throw new IllegalArgumentException("Fragment bad ID");
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/factories/FragmentFactory.kt
package ru.p3tr0vich.fuel.factories

import android.os.Bundle
import androidx.annotation.IntDef
import androidx.fragment.app.Fragment
import ru.p3tr0vich.fuel.fragments.*

object FragmentFactory {

    interface Ids {

        @Retention(AnnotationRetention.SOURCE)
        @IntDef(BAD_ID, FUELING, CALC, CHART_COST, BACKUP, PREFERENCES, ABOUT)
        annotation class Id

        companion object {
            const val BAD_ID = -1
            const val FUELING = 0
            const val CALC = 1
            const val CHART_COST = 2
            const val BACKUP = 3
            const val PREFERENCES = 4
            const val ABOUT = 5

            const val MAIN = FUELING
        }
    }

    @JvmStatic
    @JvmOverloads
    fun getFragmentNewInstance(@Ids.Id fragmentId: Int, args: Bundle? = null): Fragment {
        val fragment: Fragment = when (fragmentId) {
            Ids.CALC -> FragmentCalc()
            Ids.CHART_COST -> FragmentChartCost()
            Ids.PREFERENCES -> FragmentPreferences()
            Ids.BACKUP -> FragmentBackup()
            Ids.ABOUT -> FragmentAbout()
            Ids.FUELING -> FragmentFueling()
            Ids.BAD_ID -> throw IllegalArgumentException("Fragment bad ID")
            else -> throw IllegalArgumentException("Fragment bad ID")
        }

        fragment.arguments = args

        return fragment
    }

    @JvmStatic
    fun fragmentIdToTag(@Ids.Id id: Int): String {
        when (id) {
            Ids.ABOUT, Ids.BACKUP, Ids.CALC, Ids.CHART_COST, Ids.FUELING, Ids.PREFERENCES -> return "TAG_" + id.toString()
            Ids.BAD_ID -> throw IllegalArgumentException("Fragment bad ID")
            else -> throw IllegalArgumentException("Fragment bad ID")
        }
    }
}